rust/datafusion/src/optimizer/limit_push_down.rs

```diff
@@ -0,0 +1,164 @@
+                    projected_schema: projected_schema.clone(),
+                })
+            } else {
+                Ok(plan.clone())
+            }
+        }
+        LogicalPlan::Projection {
+            expr,
+            input,
+            schema,
+        } => {
+            // Push down limit directly (projection doesn't change number of rows)
+            Ok(LogicalPlan::Projection {
+                expr: expr.clone(),
+                input: Arc::new(limit_push_down(upper_limit, input.as_ref())?),
+                schema: schema.clone(),
+            })
+        }
+        // For other nodes we can't push down the limit
+        // But try to recurse and find other limit nodes to push down
+        _ => {
+            let expr = plan.expressions();
+
+            // apply the optimization to all inputs of the plan
+            let inputs = plan.inputs();
+            let new_inputs = inputs
+                .iter()
+                .map(|plan| limit_push_down(None, plan))
+                .collect::<Result<Vec<_>>>()?;
+
+            utils::from_plan(plan, &expr, &new_inputs)
+        }
+    }
+}
+
+impl OptimizerRule for LimitPushDown {
+    fn optimize(&self, plan: &LogicalPlan) -> Result<LogicalPlan> {
+        limit_push_down(None, plan)
+    }
+
+    fn name(&self) -> &str {
+        "limit_push_down"
+    }
+}
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        logical_plan::{col, LogicalPlan, LogicalPlanBuilder},
+        test::*,
+    };
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let rule = LimitPushDown::new();
+        let optimized_plan = rule.optimize(plan).expect("failed to optimize plan");
+        let formatted_plan = format!("{:?}", optimized_plan);
+        assert_eq!(formatted_plan, expected);
+    }
+
+    #[test]
+    fn limit_pushdown_projection_table_provider() -> Result<()> {
+        let table_scan = test_table_scan()?;
+
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .project(&[col("a")])?
+            .limit(1000)?
+            .build()?;
+
+        // Should push the limit down to table provider
+        // When it has a select
+        let expected = "Limit: 1000\
+        \n  Projection: #a\
+        \n    TableScan: test projection=None, limit=1000";
+
+        assert_optimized_plan_eq(&plan, expected);
+
+        Ok(())
+    }
+    #[test]
+    fn limit_push_down_take_smaller_limit() -> Result<()> {
+        let table_scan = test_table_scan()?;
+
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .limit(1000)?
+            .limit(10)?
+            .build()?;
+
+        // Should push down the smallest limit
+        // Towards table scan
+        // This rule doesn't replace multiple limits
+        let expected = "Limit: 10\
+        \n  Limit: 10\
+        \n    TableScan: test projection=None, limit=10";
+
+        assert_optimized_plan_eq(&plan, expected);
+
+        Ok(())
```

Comment:
I recommend two more tests:

1. Showing that the limit is not pushed through nodes where it is not correct (e.g. a Group)

```
       let plan = LogicalPlanBuilder::from(&table_scan)
            .group(....)
            .limit(1000)?

            .build()?;
```

As well one that has a multi stage limit (to ensure the logic to push later is working correctly). Perhaps

```
       let plan = LogicalPlanBuilder::from(&table_scan)
            .limit(1000)?
            .group(....)
            .limit(10)?
            .build()?;
```

Showing that the limit 1000 is pushed to the scan, but the limit 10 is not

Also, if you add support for Filter a test for that would be cool!